prov/efa/docs/efa_rdm_protocol_v4.md

```diff
@@ -0,0 +1,1408 @@
+   2. only send/receive eager messages.
+   3. does not use tagged send.
+   4. does not require FI_DIRECTED_RECV (the ability to receive only from certain address).
+
+it should be possible to receive data directly using application buffer directly. Because under such condition,
+receiver does not have special requirement on the data it is going to receive, thus will accept any message from
+the sender.
+
+However, there is one more hurdle to overcome in implementing "zero copy receive", which is the packet header
+length.
+
+Under the condition that endpoints "will only send eager messages" and "do not use tag matching", a sender will
+send data in an EAGER_MSGRTM packet. An EAGER_MSGRTM packet is consisted of packet header and application data.
+However, we cannot put packet header in application's receive buffer. Therefore, for "zero copy receive" to work,
+the receiver need to:
+
+   a. be able to estimate the packer header length of an incoming EAGER_MSGRTM packet, and the
+   b. the packet header length of EAGER_MSGRTM cannot change throughout the communication.
+
+However, there is no guarantee in the base protocol that the packet header length of EAGER_MSGRTM will not
+change.
+
+In fact, because of the existance of the handshake sub-protocol, the packet header length of an EAGER_MSGRTM
+will definitely change. Recall handshake sub-protocol's workflow is:
+
+Before receiving handshake packet, an endpoint will always include the optional raw address header in REQ packets.
+
+After receiving handshake packet, an endpoint will stop including the optional raw address header in REQ packets.
+
+The extra feature "keep packet header length stable" (stable header length) is designed to solve this problem.
+
+When an endpoint toggle on this extra request, its peer will try to satisfy it by keep the header length stable.
+Exactly how to achieve that is up to the implementation to decide, the easiest way to do that is to keep including
+raw address header in the EAGER_MSGRTM even after receiving the handshake packet.
+
+Note, because this is an extra request, an endpoint cannot assume its peer will comply with the request. Therefore,
+the receiving endpoint must be able to handle the situation that a received packet does not have the expected header length.
+
+In that case, implementation will have two choices:
+
+1. write a truncated error completion entry
+2. move the application data to right place.
+
+Note this extra request was initially introduced as an extra feature named "zero copy receive", but it is later realized
+that is not an feature because the peer does not do anything different. Rather, it is an expectation the receiving
+endpoint has for the sender. Therefore, it was re-interpreted as an extra request named "stale header length".
+This re-interpretation does not change the implementation, thus does not cause backward incompatibility.
+
+### 4.4 have connection ID in packet header (connid header)
```

Comment:
remove "have" -- "Connection ID in packet header" is descriptive enough

Reply:
Removed.

Reply:
fixed in new revision.